### PR TITLE
[Issue 16219] makeArray does unnecessary allocations for ranges with length

### PR DESCRIPTION
**Pull request by JackStouffer**

### CODE REVIEW

**Comment by wilzbach** [comment]
Looks reasonable too me :)

File: std/experimental/allocator/package.d [hunk 1]
```diff
@@ -737,9 +737,13 @@ unittest
 T[] makeArray(T, Allocator, R)(auto ref Allocator alloc, R range)
 if (isInputRange!R)
 {
-    static if (isForwardRange!R)
+    static if (isForwardRange!R || hasLength!R)
     {
-        size_t length = walkLength(range.save);
+        static if (hasLength!R)
+            size_t length = range.length;
```
Comment (Dgame) [inline std/experimental/allocator/package.d]: `immutable`?

**Comment by JackStouffer** [comment]
@Dgame Fixed

**Comment by ZombineDev** [comment]
LGTM DbI FTW

**Comment by burner** [comment]
LGTM

**Comment by burner** [comment]
Auto-merge toggled on

**Pull request merged**